### PR TITLE
[Review] add security policy Basic256Sha256 support and PKI with OpenSSL provider

### PR DESCRIPTION
**Pull request by wenshan1**
1. I introduce two new macros: UA_ENABLE_ENCRYPTION_MBEDTLS UA_ENABLE_ENCRYPTION_OPENSSL.
The encryption code is divided into two parts: the encryption common code and the concrete encryption algorithm implementation code.
The encryption common code includes:
src/ua_securechannel.c:415:#ifdef UA_ENABLE_ENCRYPTION
src/client/ua_client_connect.c:304:#ifdef UA_ENABLE_ENCRYPTION
plugins/include/open62541/plugin/pki_default.h:19:#ifdef UA_ENABLE_ENCRYPTION
...
The concrete encryption algorithm implementation code includes the source code written by mbedtls encryption provider or the source code written by opessl encryption provider.
The user of open6254 (such as: the examples) will not care who is the encryption provider, just enable the macro UA_ENABLE_ENCRYPTION to get the encryption function.
2. For the previous compatibility and do not modify the test code and examples code, UA_ENABLE_ENCRYPTION is enabled then the macro UA_ENABLE_ENCRYPTION_MBEDTLS is enabled by default.
The user can use the OpenSSL as the security provider like this: cmake .. -DUA_BUILD_EXAMPLES=ON -DUA_ENABLE_ENCRYPTION_OPENSSL=ON
3. Add security policy Basic256Sha256 support, the other security policy with OpenSSL will be implemented later.
4. Add the PKI support with OpenSSL provider.

5. Test:
OS environment: Ubuntu18.04 Linux
The PKI is enabled.
a.  The UA encryption client with MbedTLs can open the secure channel and active the session with the server with OpenSSL.
b. The opcfoundation CTT tool v1.03 as the client can open the secure channel, active the session and browse the server namespace in the server with OpenSSL.
c. The related test cases of Basic256SHA256 and PKI in the  opcfoundation CTT tool are passed.

6. The Unit test/CTest will be added later.

### CODE REVIEW

**Comment by wenshan1** [comment]
reopen to run the checks.

**Comment by jpfr** [comment]
It seems you are still working on this.
Please change the title to [Review] when we should take a look.

Can you also squash commits to get of the fix-commits?

**Comment by jpfr** [comment]
Looking very good otherwise!

**Comment by wenshan1** [comment]
Please review it and merge it into master branch if no comments. Thank you.

**Comment by wenshan1** [comment]
I always fail for check on windows OS, Really crazy.

File: plugins/ua_pki_default.c [hunk 1]
```diff
@@ -422,48 +466,6 @@ certificateVerification_verify(void *verificationContext,
     return retval;
 }
 
-/* Find binary substring. Taken and adjusted from
- * http://tungchingkai.blogspot.com/2011/07/binary-strstr.html */
-
-static const unsigned char *
-bstrchr(const unsigned char *s, const unsigned char ch, size_t l) {
```
Comment (wenshan1) [inline plugins/ua_pki_default.c]: Move the function bstrchr() out of the macro UA_ENABLE_ENCRYPTION_MBEDTLS. So it can be re-used.

**Comment by jpfr** [comment]
It could be a general problem with some dependencies.
We will look into it.

**Comment by wenshan1** [comment]
No new source code added, only trigger the CI checks by "git push -f".

**Comment by jpfr** [comment]
Rebased to fix the CI issues.

**Comment by jpfr** [comment]
Very nice! Merged.

**Pull request merged**